### PR TITLE
[zero] refactor memstats_collector

### PR DESCRIPTION
**Pull request by 1SAA**

### CODE REVIEW

File: colossalai/utils/memory_tracer/memstats_collector.py [hunk 1]
```diff
@@ -31,57 +31,34 @@ def __init__(self) -> None:
...
-            return [elem / scale for elem in self._non_model_data_cuda_list]
+            return self._non_model_data_cuda_list
         elif device_type == 'cpu':
-            return [elem / scale for elem in self._non_model_data_cpu_list]
+            return self._non_model_data_cpu_list
         else:
             raise TypeError
 
-    def max_non_model_data(self, device_type: str) -> int:
+    def next_max_non_model_data(self, device_type: str) -> int:
```
Comment (feifeibear) [inline colossalai/utils/memory_tracer/memstats_collector.py:61]: next_period_non_model_data?
That is more straitforwards

Comment (feifeibear) [reply colossalai/utils/memory_tracer/memstats_collector.py:61]: or next_period_non_model_data_usage?

File: colossalai/zero/utils/zero_hook.py [hunk 2]
```diff
@@ -77,11 +54,10 @@ def post_fwd_exec(self, module: torch.nn.Module, *args):
                 tensor_list.append(param.colo_attr.sharded_data_tensor)
             self.shard_strategy.shard(tensor_list, self.process_group)
 
-        # remove torch payload
-        for param in module.parameters(recurse=False):
-            param.colo_attr.set_data_none()
-
-    def pre_bwd_exec(self, module: torch.nn.Module, input, output):
+    def payload_movement(self, module: torch.nn.Module):
```
Comment (feifeibear) [inline colossalai/zero/utils/zero_hook.py:57]: The name is bad. Actually, you did tensor moving, tensor state changing and mem sampling.

File: colossalai/zero/utils/zero_hook.py [hunk 3]
```diff
@@ -92,18 +68,32 @@ def pre_bwd_exec(self, module: torch.nn.Module, input, output):
             for param in module.parameters(recurse=False):
```
Comment (ver217) [inline colossalai/zero/utils/zero_hook.py:68]: You can remove this branch now, since `self._stateful_tensor_mgr` can't be `None`.

File: colossalai/zero/utils/zero_hook.py [hunk 4]
```diff
@@ -92,18 +68,32 @@ def pre_bwd_exec(self, module: torch.nn.Module, input, output):
...
-            for param in module.parameters(recurse=False):
-                assert hasattr(param, 'colo_attr')
-                tensor_list.append(param.colo_attr.sharded_data_tensor)
-            self.shard_strategy.gather(tensor_list, self.process_group)
-
-        # record memory statistics
+        # record model data statistics
         if self._memstarts_collector:
-            self._memstarts_collector.sample_memstats()
+            self._memstarts_collector.sample_model_data()
```
Comment (feifeibear) [inline colossalai/zero/utils/zero_hook.py:73]: why sample model data here?

Comment (feifeibear) [reply colossalai/zero/utils/zero_hook.py:73]: I think model data sampling should be done at the same time of overall data sampling.

File: colossalai/zero/utils/zero_hook.py (hunk 3 above)
Comment (1SAA) [reply colossalai/zero/utils/zero_hook.py:68]: Sure.

File: colossalai/zero/utils/zero_hook.py (hunk 2 above)
Comment (1SAA) [reply colossalai/zero/utils/zero_hook.py:57]: May be `adjust_model_data`?

File: colossalai/utils/memory_tracer/memstats_collector.py (hunk 1 above)
Comment (1SAA) [reply colossalai/utils/memory_tracer/memstats_collector.py:61]: Is `future_non_model_data_usage` ok?

File: colossalai/utils/memory_tracer/memstats_collector.py [hunk 5]
```diff
@@ -107,9 +82,37 @@ def start_collection(self):
...
+            self._overall_cpu_list.append(colo_device_memory_used(torch.device('cpu')))
+
+            assert len(self._model_data_cuda_list) == len(self._overall_cuda_list)
+
+            self._non_model_data_cuda_list.append(self._overall_cuda_list[-1] - self._model_data_cuda_list[-1])
+            self._non_model_data_cpu_list.append(self._overall_cpu_list[-1] - self._model_data_cpu_list[-1])
+            self._sampling_time.append(time.time())
+            self._mem_monitor.start()
+
     def sample_memstats(self) -> None:
```
Comment (ver217) [inline colossalai/utils/memory_tracer/memstats_collector.py:116]: Where do you call this method?

Comment (1SAA) [reply colossalai/utils/memory_tracer/memstats_collector.py:116]: Currently, test_stateful_tensor_mgr.py.
I will change the code in test_stateful_tensor_mgr.py soon.
So it will appear in nowhere.
But I just keep it for some recording operations.

File: colossalai/zero/utils/tensor_placement_policy.py [hunk 6]
```diff
@@ -70,7 +70,7 @@ def evict_tensors(self,
             freed_cuda_model_data = 0
             to_free_tensor_list = hold_cuda_tensor_list
             if not warmup:
-                next_compute_idx: Dict[StatefulTensor, int] = {t: len(compute_list) for t in hold_cuda_tensor_list}
```
Comment (ver217) [inline colossalai/zero/utils/tensor_placement_policy.py:73]: Why did you delete the type hint?

Comment (1SAA) [reply colossalai/zero/utils/tensor_placement_policy.py:73]: Because `next_compute_idx` become List[StatefulTensor, int] below.
This type hint is not consistent.

File: colossalai/zero/utils/zero_hook.py (hunk 2 above)
Comment (feifeibear) [reply colossalai/zero/utils/zero_hook.py]: that looks better

File: colossalai/zero/utils/zero_hook.py [hunk 7]
```diff
@@ -55,10 +45,33 @@ def pre_fwd_exec(self, module: torch.nn.Module, *args):
...
+    def shard_parameters(self, module: torch.nn.Module):
+        # shard gathered parameters
+        if module.param_is_sharded:
+            tensor_list = []
+            for param in module.parameters(recurse=False):
+                assert hasattr(param, 'colo_attr')
+                tensor_list.append(param.colo_attr.sharded_data_tensor)
+            self.shard_strategy.shard(tensor_list, self.process_group)
+
+    def adjust_module_param(self, module: torch.nn.Module):
```
Comment (feifeibear) [inline colossalai/zero/utils/zero_hook.py:57]: adjust_model_tenors_and_sampling?

File: colossalai/utils/memory_tracer/memstats_collector.py (hunk 1 above)
Comment (1SAA) [reply colossalai/utils/memory_tracer/memstats_collector.py]: done

**Review by feifeibear — APPROVED** [review]
